cmd/snap/inhibit.go

```diff
@@ -0,0 +1,152 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"time"
+
+	"github.com/snapcore/snapd/cmd/snaplock/runinhibit"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/progress"
+)
+
+func inhibitMessage(snapName string, hint runinhibit.Hint) string {
+	switch hint {
+	case runinhibit.HintInhibitedForRefresh:
+		return fmt.Sprintf(i18n.G("snap package %q is being refreshed, please wait"), snapName)
+	default:
+		return fmt.Sprintf(i18n.G("snap package cannot be used now: %s"), string(hint))
+	}
+}
+
+func isGraphicalSession() bool {
+	return os.Getenv("DISPLAY") != "" || os.Getenv("WAYLAND_DISPLAY") != ""
+}
+
+var hasZenityExecutable = func() bool {
+	return osutil.ExecutableExists("zenity")
+}
+
+func zenityFlow(snapName string, hint runinhibit.Hint) error {
+	zenityTitle := i18n.G("Snap package waiting for update")
+
+	// Run zenity with a progress bar.
+	// TODO: while we are waiting ask snapd for progress updates and send those
+	// to zenity via stdin.
+	zenityDied := make(chan error, 1)
+
+	// TODO: use a dbus API to allow integration with native desktop environment.
+	cmd := exec.Command(
+		"zenity",
+		// [generic options]
+		"--title="+zenityTitle,
+		// [progress options]
+		"--progress",
+		"--text="+inhibitMessage(snapName, hint),
+		"--pulsate",
+	)
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	// Make sure that zenity is eventually terminated.
+	defer cmd.Process.Signal(os.Interrupt)
+	// Wait for zenity to terminate and store the error code.
+	// The way we invoke zenity --progress makes it wait forever.
+	// so it will typically be an external operation.
+	go func() {
+		zenityErr := cmd.Wait()
+		if zenityErr != nil {
+			zenityErr = fmt.Errorf("zenity error: %s\n", zenityErr)
+		}
+		zenityDied <- zenityErr
+	}()
+
+	if err := waitInhibitUnlock(snapName, runinhibit.HintNotInhibited, zenityDied); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func textFlow(snapName string, hint runinhibit.Hint) error {
+	fmt.Fprintf(Stdout, "%s\n", inhibitMessage(snapName, hint))
+	pb := progress.MakeProgressBar()
+	pb.Spin(i18n.G("please wait..."))
+	err := waitInhibitUnlock(snapName, runinhibit.HintNotInhibited, nil)
+	pb.Finished()
+	return err
+}
+
+func waitWhileInhibited(snapName string) error {
+	hint, err := runinhibit.IsLocked(snapName)
+	if err != nil {
+		return err
+	}
+	if hint == runinhibit.HintNotInhibited {
+		return nil
+	}
+
+	// wait for HintInhibitedForRefresh set by gate-auto-refresh hook handler
+	// when it has finished; the hook starts with HintInhibitedGateRefresh lock
+	// and then either unlocks it or changes to HintInhibitedForRefresh (see
+	// gateAutoRefreshHookHandler in hooks.go).
+	// waitInhibitUnlock will return also on HintNotInhibited.
+	if err := waitInhibitUnlock(snapName, runinhibit.HintInhibitedForRefresh, nil); err != nil {
+		return err
+	}
+
```

Comment:
at this point the lock could be already again runinhibit.HintNotInhibited  if the hook didn't block things, so we shouldn't display the zenity or text UI in that case but return like we do in 106, so it seems waitInhibitUnlock should return whether it hit no inhibit or a different waitFor condition

Reply:
Done.